src/imitation/algorithms/preference_comparisons.py

```diff
@@ -0,0 +1,289 @@
+"""Learning reward models using preference comparisons.
+
+Trains a reward model and optionally a policy based on preferences
+between short trajectory fragments.
+"""
+
+
+import abc
+from typing import Callable, List, Optional, Tuple
+
+import numpy as np
+import torch as th
+
+from imitation.data.fragments import Fragmenter, RandomFragmenter
+from imitation.data.rollout import flatten_trajectories_with_rew
+from imitation.data.types import TrajectoryWithRew, TransitionsWithRew
+from imitation.policies.trainer import AgentTrainer
+from imitation.rewards.reward_nets import RewardNet
+
+PreferenceGatherer = Callable[
+    [List[Tuple[TrajectoryWithRew, TrajectoryWithRew]]], np.ndarray
+]
+"""Gathers the probabilities that fragment 1 is preferred for a batch of fragments.
+
+Takes a list of pairs of trajectory fragments as input. Should return a numpy
+array with shape (b, ), where b is the length of the list (i.e. batch size).
+Each item in the array is the probability that fragment 1 is preferred over
+fragment 2 for the corresponding pair of fragments.
+
+Note that for human feedback, these probabilities are simply 0 or 1
+(or 0.5 in case of indifference), but synthetic models may yield other
+probabilities.
+"""
+
+
+class SyntheticGatherer:
+    """Computes synthetic preferences using ground-truth environment rewards."""
+
+    def __init__(self, probabilistic: bool = False, noise_prob: float = 0.0):
+        """Initialize the synthetic preference gatherer.
+
+        Args:
+            probabilistic: if False (default), the preferences are either 1
+                (if fragment 1 has a higher sum of rewards), 0 (if fragment 2
+                has a higher sum of rewards) or 0.5 (if they are equal).
+                If True, a probability between 0 and 1 is returned
```

Comment:
I think I also like your suggestion better, especially because it's closer to human feedback and slightly less trivial. In principle, I also like a temperature argument because I think it could replace both `probabilistic` and `noise_prob` (an infinite temperature would be equivalent to `noise_prob=1`). OTOH, the DRLHP paper as well as my own default implementation of the loss function assume the `noise_prob` version to model human error, rather than a temperature. It might be nice to have this be the same everywhere, and I don't really like having both `noise_prob` and `temperature`. Do you have a preference?